server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseIT.java

```diff
@@ -494,4 +492,124 @@ public void onFailure(final Exception e) {
+            final long initialRetainingSequenceNumber,
+            final BiConsumer<IndexShard, ActionListener<ReplicationResponse>> indexShard,
+            final Consumer<IndexShard> afterSync) throws InterruptedException {
+        final Settings settings = Settings.builder()
+                .put("index.number_of_shards", 1)
+                .put("index.number_of_replicas", internalCluster().numDataNodes())
+                .put(IndexService.RETENTION_LEASE_SYNC_INTERVAL_SETTING.getKey(), TimeValue.timeValueSeconds(1))
+                .build();
+        assertAcked(prepareCreate("index").setSettings(settings));
+        ensureYellowAndNoInitializingShards("index");
+        assertFalse(client().admin().cluster()
+            .prepareHealth("index").setWaitForActiveShards(internalCluster().numDataNodes()).get().isTimedOut());
+
+        final String primaryShardNodeId = clusterService().state().routingTable().index("index").shard(0).primaryShard().currentNodeId();
+        final String primaryShardNodeName = clusterService().state().nodes().get(primaryShardNodeId).getName();
+        final IndexShard primary = internalCluster()
+                .getInstance(IndicesService.class, primaryShardNodeName)
+                .getShardOrNull(new ShardId(resolveIndex("index"), 0));
+
+        final String source = randomAlphaOfLength(8);
+        final CountDownLatch latch = new CountDownLatch(1);
+        final ActionListener<ReplicationResponse> listener = ActionListener.wrap(r -> latch.countDown(), e -> fail(e.toString()));
+        primary.addRetentionLease(idForInitialRetentionLease, initialRetainingSequenceNumber, source, listener);
+        latch.await();
+
+        final String waitForActiveValue = randomBoolean() ? "all" : Integer.toString(internalCluster().numDataNodes() + 1);
+
+        client()
+                .admin()
+                .indices()
+                .prepareUpdateSettings("index")
+                .setSettings(Settings.builder().put("index.write.wait_for_active_shards", waitForActiveValue).build())
+                .get();
+
+        try {
+            final CountDownLatch actionLatch = new CountDownLatch(1);
+            final AtomicBoolean success = new AtomicBoolean();
+
+            indexShard.accept(
+                    primary,
+                    new ActionListener<ReplicationResponse>() {
+
+                        @Override
+                        public void onResponse(final ReplicationResponse replicationResponse) {
+                            success.set(true);
+                            actionLatch.countDown();
+                        }
+
+                        @Override
+                        public void onFailure(final Exception e) {
+                            fail(e.toString());
+                        }
+
+                    });
+            actionLatch.await();
+            assertTrue(success.get());
+            afterSync.accept(primary);
+        } finally {
+            client()
```

Comment:
This finally block is not needed. The reason that we have it in the run under blocks test is because we can't delete the index in test cleanup if there is a write block in place. 😄